qos-monitor/src/main/java/eu/arrowhead/core/qos/service/ping/monitor/impl/DefaultExternalPingMonitor.java

```diff
@@ -0,0 +1,204 @@
+		logger.debug("validateAcknowledgedMeasurmentRequest started...");
+
+		try {
+			Assert.notNull(acknowledgedMeasurmentRequest, "IcmpPingRequestACK is null");
+			Assert.notNull(acknowledgedMeasurmentRequest.getAckOk(), "IcmpPingRequestACK.ackOk is null");
+			Assert.isTrue(acknowledgedMeasurmentRequest.getAckOk().equalsIgnoreCase("OK"), "IcmpPingRequestACK is null");
+
+		} catch (final Exception ex) {
+			logger.warn("External pingMonitorProvider replied invalid ack : " + ex);
+
+			throw new ArrowheadException("External pingMonitorProvider replied invalid ack", ex);
+		}
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	private IcmpPingRequestDTO createIcmpPingRequest(final String address) {
+		logger.debug("createIcmpPingRequest started...");
+
+		final IcmpPingRequestDTO request = new IcmpPingRequestDTO();
+		request.setHost(address);
+		request.setPacketSize(pingMeasurementProperties.getPacketSize());
+		request.setTimeout(Long.valueOf(pingMeasurementProperties.getTimeout()));
+		request.setTimeToRepeat(pingMeasurementProperties.getTimeToRepeat());
+		request.setTtl(ICMP_TTL);
+
+		return request;
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	private int calculateTimeOut() {
+
+		final int singlePingTimeOut = pingMeasurementProperties.getTimeout();
+		final int timesToRepeatPing = pingMeasurementProperties.getTimeToRepeat();
+
+		return singlePingTimeOut * timesToRepeatPing * OVERHEAD_MULTIPLIER;
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	private UriComponents createPingMonitorProviderUri() {
+		logger.debug("createPingMonitorProviderUri started...");
+
+		return Utilities.createURI(pingMonitorSecure ? CommonConstants.HTTPS : CommonConstants.HTTP, externalPingMonitorAddress, externalPingMonitorPort, externalPingMonitorPath);
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	private UriComponents createPingMonitorProviderEchoUri() {
+		logger.debug("createPingMonitorProviderEchoUri started...");
+
+		return Utilities.createURI(pingMonitorSecure ? CommonConstants.HTTPS : CommonConstants.HTTP, externalPingMonitorAddress, externalPingMonitorPort, CommonConstants.ECHO_URI);
+	}
+
+	//-------------------------------------------------------------------------------------------------
+	private SystemRequestDTO getPingMonitorSystemRequestDTO() {
+		logger.debug("getPingMonitorSystemRequestDTO started...");
+
+		final SystemRequestDTO system = new SystemRequestDTO();
+		system.setSystemName(externalPingMonitorName);
+		system.setAddress(externalPingMonitorAddress);
+		system.setPort(externalPingMonitorPort);
+		system.setMetadata(null);
+		if (pingMonitorSecure) {
+			system.setAuthenticationInfo(externalPingMonitorAuthInfo);
+		}
+
+		return system;
+	}
```

Comment:
should be calculated once (at first call), then stored in a member, and all later calls should just return the stored member.

Reply:
done